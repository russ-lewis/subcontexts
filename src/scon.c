#include <stdio.h>
#include <stdlib.h>
#include <dlfcn.h>
#include <unistd.h>
#include <limits.h>
#include <string.h>

#include <mem/map.h>
#include <scon.h>

void scon_init(scon_t *self, const char *libpath)
{
    // dlopen does not follow relative paths, so find the absolute path of libpath
    char abspath[PATH_MAX];
    if (libpath[0] == '/')
    {
        memcpy(abspath, libpath, strlen(libpath));
    }
    else
    {
        getcwd(abspath, sizeof(abspath));
        strcat(abspath, "/");
        strcat(abspath, libpath);
    }

    Map *before = Map_parse(-1);

    // open the library
    self->libhandle = dlopen(abspath, RTLD_NOW);
    if (self->libhandle == NULL)
    {
        fprintf(stderr, dlerror());
        exit(1);
    }
    
    Map *after = Map_parse(-1);
    Map *diff = Map_diff(before, after);
    Map_free(before); 
    Map_free(after);

    if (diff->head != NULL)
    {
        for (MapEntry *curr = diff->head; curr != NULL; curr = curr->next)
        {
            // TODO need to copy the pages?
        }
    }    
    
    Map_free(diff);
}

void *scon_loadf(scon_t *self, const char *funcname)
{
    void *func = dlsym(self->libhandle, funcname);
    if (func == NULL)
    {
        fprintf(stderr, dlerror());
    }

    return func;
}

void scon_free(scon_t *self)
{
    dlclose(self->libhandle);
}

