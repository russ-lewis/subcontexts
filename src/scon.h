/**
 * Includes functions to initialize and uses subcontexts.
 */

#ifndef SCON_H
#define SCON_H

#include <stdio.h>
#include <mem/map.h>

typedef unsigned int scon_t;

/**
 * Initializes a subcontext given the path to the library.
 */
scon_t scon_create(const char *);

/**
 * Calls a function in the library.
 */
void *scon_callf(const scon_t, const char *, void *);

/**
 * Frees resources held by the subcontext.
 */
void scon_close(const scon_t);

#endif
