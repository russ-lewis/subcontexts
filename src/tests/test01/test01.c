#include <stdio.h>
#include <stdlib.h>
#include <dlfcn.h>
#include <scon.h>

int main()
{
    scon_t scon = scon_create("test01_scon.so"); 
    scon_callf(scon, "test", NULL);
    scon_close(scon);
    return 0;
}
