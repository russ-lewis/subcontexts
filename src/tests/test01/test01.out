in test01_scon
